haskell/Java/HaskellBaseLexer.java

```diff
@@ -0,0 +1,367 @@
+        setChannel(HIDDEN);
+        if (pendingDent) {
+            indentCount += getText().length();
+        }
+    }
+
+    private int getSavedIndent() { return indentStack.empty() ? startIndent : indentStack.peek().second(); }
+
+    private CommonToken
+    createToken(int type, String text, Token next) {
+        CommonToken token = new CommonToken(type, text);
+        if (initialIndentToken != null) {
+            token.setStartIndex(initialIndentToken.getStartIndex());
+            token.setLine(initialIndentToken.getLine());
+            token.setCharPositionInLine(initialIndentToken.getCharPositionInLine());
+            token.setStopIndex(next.getStartIndex() - 1);
+        }
+        return token;
+    }
+
+    private void processINToken(Token next) {
+        while (!indentStack.empty() && indentStack.peek().first() != "let") {
+            tokenQueue.offer(createToken(HaskellLexer.SEMI, "SEMI", next));
+            tokenQueue.offer(createToken(HaskellLexer.VCCURLY, "VCCURLY", next));
+            nestedLevel--;
+            indentStack.pop();
+        }
+
+        if (!indentStack.empty() && indentStack.peek().first().equals("let")) {
+            tokenQueue.offer(createToken(HaskellLexer.SEMI, "SEMI", next));
+            tokenQueue.offer(createToken(HaskellLexer.VCCURLY, "VCCURLY", next));
+            nestedLevel--;
+            indentStack.pop();
+        }
+    }
+
+    private void processEOFToken(Token next) {
+        indentCount = startIndent;
+        if (!pendingDent) {
+            initialIndentToken = next;
+        }
+
+        while (nestedLevel > indentStack.size()) {
+            if (nestedLevel > 0)
+                nestedLevel--;
+
+            tokenQueue.offer(createToken(HaskellLexer.SEMI, "SEMI", next));
+            tokenQueue.offer(createToken(HaskellLexer.VCCURLY, "VCCURLY", next));
+        }
+
+        while (indentCount < getSavedIndent()) {
+            if (!indentStack.empty() && nestedLevel > 0) {
+                indentStack.pop();
+                nestedLevel--;
+            }
+
+            tokenQueue.offer(createToken(HaskellLexer.SEMI, "SEMI", next));
+            tokenQueue.offer(createToken(HaskellLexer.VCCURLY, "VCCURLY", next));
+        }
+
+        if (indentCount == getSavedIndent()) {
+            tokenQueue.offer(createToken(HaskellLexer.SEMI, "SEMI", next));
+        }
+
+        if (wasModuleExport) {
+            tokenQueue.offer(createToken(HaskellLexer.VCCURLY, "VCCURLY", next));
+        }
+
+        startIndent = -1;
+    }
+
+    // Algorithm's description here:
+    // https://www.haskell.org/onlinereport/haskell2010/haskellch10.html
+    // https://en.wikibooks.org/wiki/Haskell/Indentation
+    @Override
+    public Token nextToken() {
+        if (!tokenQueue.isEmpty()) {
+            return tokenQueue.poll();
+        }
+
+        Token next = super.nextToken();
+
+        if (startIndent == -1
+            && next.getType() != HaskellLexer.NEWLINE
```

Comment:
Maybe it makes sense to calculate this only once and use it in other places? `type = next.getType()`

Reply:
Fixed